# Socrata API Project - Nashville.gov Top 500 Monthly Searches

In [1]:
import requests
import matplotlib.pyplot as plt

### Q1. Make an API request that returns the months where "fire" was searched in 2016. Which month had the most searches?

In [2]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?year=2016&query_text=fire'
response = requests.get(endpoint)

In [3]:
response

<Response [200]>

In [4]:
res_2016_fire = response.json()
res_2016_fire

[{'month_name': 'January',
  'year': '2016',
  'query_count': '19',
  'query_text': 'fire'},
 {'month_name': 'February',
  'year': '2016',
  'query_count': '35',
  'query_text': 'fire'},
 {'month_name': 'March',
  'year': '2016',
  'query_count': '32',
  'query_text': 'fire'},
 {'month_name': 'April',
  'year': '2016',
  'query_count': '26',
  'query_text': 'fire'},
 {'month_name': 'May',
  'year': '2016',
  'query_count': '24',
  'query_text': 'fire'},
 {'month_name': 'June',
  'year': '2016',
  'query_count': '31',
  'query_text': 'fire'},
 {'month_name': 'July',
  'year': '2016',
  'query_count': '24',
  'query_text': 'fire'},
 {'month_name': 'August',
  'year': '2016',
  'query_count': '47',
  'query_text': 'fire'},
 {'month_name': 'September',
  'year': '2016',
  'query_count': '36',
  'query_text': 'fire'},
 {'month_name': 'October',
  'year': '2016',
  'query_count': '38',
  'query_text': 'fire'},
 {'month_name': 'November',
  'year': '2016',
  'query_count': '32',
  'query_text

In [5]:
# Order results by query_count DESC to get the max query_count first
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?year=2016&query_text=fire&$order=query_count DESC'
response = requests.get(endpoint)
max_fire = response.json()
max_fire

[{'month_name': 'August',
  'year': '2016',
  'query_count': '47',
  'query_text': 'fire'},
 {'month_name': 'December',
  'year': '2016',
  'query_count': '38',
  'query_text': 'fire'},
 {'month_name': 'October',
  'year': '2016',
  'query_count': '38',
  'query_text': 'fire'},
 {'month_name': 'September',
  'year': '2016',
  'query_count': '36',
  'query_text': 'fire'},
 {'month_name': 'February',
  'year': '2016',
  'query_count': '35',
  'query_text': 'fire'},
 {'month_name': 'March',
  'year': '2016',
  'query_count': '32',
  'query_text': 'fire'},
 {'month_name': 'November',
  'year': '2016',
  'query_count': '32',
  'query_text': 'fire'},
 {'month_name': 'June',
  'year': '2016',
  'query_count': '31',
  'query_text': 'fire'},
 {'month_name': 'April',
  'year': '2016',
  'query_count': '26',
  'query_text': 'fire'},
 {'month_name': 'July',
  'year': '2016',
  'query_count': '24',
  'query_text': 'fire'},
 {'month_name': 'May',
  'year': '2016',
  'query_count': '24',
  'query_tex

In [6]:
# Limit 1 to get the month with the top query_count for 'fire'
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?year=2016&query_text=fire&$order=query_count DESC&$limit=1'
response = requests.get(endpoint)
max_fire = response.json()
max_fire

[{'month_name': 'August',
  'year': '2016',
  'query_count': '47',
  'query_text': 'fire'}]

#### A1. August had the most searches for 'fire' at 47.

### Q2. Make an API request that returns all the times a query was run more than 100 times in a month. How many times did this occur?

In [7]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?$where=query_count > 100'
response = requests.get(endpoint)
over100 = response.json()
# over100

In [8]:
times = len(over100)
times
# 1,000 looks too exact... investigating found this:
# https://stackoverflow.com/questions/30941170/why-am-i-limited-to-1-000-rows-on-soda-api-when-i-have-an-app-key
# It defaults to 1,000 records if limit is not specified

1000

In [9]:
# I can specify a limit OR use count:
# https://stackoverflow.com/questions/27823200/is-there-a-socrata-api-method-to-get-the-row-count-of-a-dataset
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?$select=count(*)&$where=query_count > 100'
response = requests.get(endpoint)
over100_count = response.json()
over100_count

[{'count': '1262'}]

#### A2. 1,262 is the number of times a query was run more than 100 times in a month.

### Q3. Make another API request that returns all the times "codes" was searched more than 100 times in a month. How many times did this occur?

In [12]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?$select=query_text&$where=query_count > 100 AND query_text=codes'
response = requests.get(endpoint)
over100_codes_count = response.json()
over100_codes_count
# Not working...

{'message': 'Query coordinator error: query.soql.no-such-column; No such column: codes; position: Map(row -> 1, column -> 66, line -> "SELECT `query_text` WHERE `query_count` > 100 AND `query_text` = `codes`\\n                                                                 ^")',
 'errorCode': 'query.soql.no-such-column',
 'data': {'column': 'codes',
  'dataset': 'alpha.6008',
  'position': {'row': 1,
   'column': 66,
   'line': 'SELECT `query_text` WHERE `query_count` > 100 AND `query_text` = `codes`\n                                                                 ^'}}}

In [13]:
# Omitting select:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?$where=query_count > 100 AND query_text=codes'
response = requests.get(endpoint)
over100_codes_count = response.json()
over100_codes_count
# Still thinks codes is a column, maybe the where statement is wrong?

{'message': 'Query coordinator error: query.soql.no-such-column; No such column: codes; position: Map(row -> 1, column -> 112, line -> "SELECT `month_name`, `month`, `year`, `query_count`, `query_text` WHERE `query_count` > 100 AND `query_text` = `codes`\\n                                                                                                               ^")',
 'errorCode': 'query.soql.no-such-column',
 'data': {'column': 'codes',
  'dataset': 'alpha.6008',
  'position': {'row': 1,
   'column': 112,
   'line': 'SELECT `month_name`, `month`, `year`, `query_count`, `query_text` WHERE `query_count` > 100 AND `query_text` = `codes`\n                                                                                                               ^'}}}

In [18]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?query_text=codes&query_count > 100'
response = requests.get(endpoint)
over100_codes_count = response.json()
over100_codes_count
# A new error message means I'm getting closer right?!

{'error': True, 'message': 'Unrecognized arguments [query_count > 100]'}

In [20]:
endpoint = 'https://data.nashville.gov/resource/fuaa-r5cm.json?query_text=codes&$where=query_count > 100'
response = requests.get(endpoint)
over100_codes_count = response.json()
len(over100_codes_count)
# Hope this is right

56

#### A3. 56 is the number of times "codes" was searched more than 100 times in a month.

### Q4. Make an API request that returns the entire Top 500 Monthly Searches dataset. Make a chart that shows the number of times "maps" was searched in a month across the entire time frame.